### PR TITLE
Adds support for GitHub Actions

### PR DESCRIPTION
**Pull request by marceltaeumel**
Here is an example on what it will look like:
https://github.com/marceltaeumel/squeak-app/actions/runs/1262385209

Details:
- Drops support for TravisCI
- Drops support for Squeak 5.0 and 5.1
- Splits up prepare.sh into prepare_image.sh, test_image.sh, prepare_bundles.sh, and deploy_bundles.sh
- Extracts prepare_image_post.sh from (original) prepare_image.sh to be used during bundle creation (such as for collecting the extra Etoys project files)
- For testing, patches smalltalkCI to support GHA via gha-support.cs in test_image.st (can be removed later; see https://github.com/hpi-swa/smalltalkCI/issues/541)
- Uses Windows platform to prepare both 32-bit and 64-bit images, and to test the images
- Uses macOS platform to create (and sign/notarize) and deploy the bundles
- All intermediate build artifacts will be uploaded during the action to remain inspectable for 90 days; this does not replace the deployment step to files.squeak.org

Next steps:
- `bundle.yml`: Update `push:` filter to only work on "squeak-trunk"
- `prepare_bundles.sh`: Provide secrets to prepare signing and ssh/rsync
- `prepare_aio.sh`: Provide secrets to sign and notarize the all-in-one for macOS
- `prepare_mac.sh`: Provide secrets to sign and notarize the all-in-one for macOS
- `deploy_bundles.sh`: Provide secrets for rsync
- Merge changes back to "squeak-5.1" and "squeak-5.2" branches, then updating build matrix, push filter, and DEPLOYMENT_BRANCH in each branch
- Enable mail delivery to `squeak-dev@lists.squeakfoundation.org`

I will take a look at sign/notarize soon. Maybe one of you is more experienced in this field already and can thus help out. :-)

### CODE REVIEW

File: .github/workflows/bundle.yml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      # - squeak-5.2
+    paths-ignore:
+      - '**.md' # Skip changes in documentation artifacts
+  pull_request:
+    branches:
+      - squeak-trunk
+      # - squeak-5.3
+      # - squeak-5.2
+    paths-ignore:
+      - '**.md'
```
Comment (LinqLover) [inline .github/workflows/bundle.yml]: If we build on pushes to every branch, why do we need double builds for every PR then?

File: .github/workflows/bundle.yml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.smalltalk }}
+          path: |
+            tmp/*.sources
+            tmp/*.image
+            tmp/*.changes
+            tmp/version.sh
+            !tmp/Test*
```
Comment (LinqLover) [inline .github/workflows/bundle.yml]: Just out of interest, where is the 90-day limit you mention indicated? :)

File: env_vars [hunk 3]
```diff
@@ -0,0 +1,14 @@
+readonly HOME_DIR="$(pwd)"
```
Comment (LinqLover) [inline env_vars]: If you named this file with a `.env` suffix, editors such as VS Code could display them more conveniently, I think. :-)

File: env_vars [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+readonly TMP_DIR="${HOME_DIR}/tmp"
+
+readonly FILES_BASE="http://files.squeak.org/base"
+readonly RELEASE_URL="${FILES_BASE}/${SMALLTALK_VERSION/Etoys/Squeak}"
```
Comment (LinqLover) [inline env_vars:4]: Why is there an `Etoys`?

**Review by LinqLover — COMMENTED** [review]
Not a review, even though GitHub will label it like one, just some curious questions. :-)

Can we also address [this issue](https://github.com/squeak-smalltalk/squeak-app/pull/17#issuecomment-877329529) (duplicate `squeak.sh` file) or should this go into a separate PR?

As part of this change, could we also consider replacing the recurring "Hi all!" commit by a GitHub action that sends the failure mail to the list (something like [this](https://github.com/LinqLover/TelegramBot/blob/098a29bf88037f6e50f2c0db60a403946d8dad14/.github/workflows/main.yml#L43-L54))? I think this would make the commit history look cleaner and it's also more transparent. :-)

File: .github/workflows/bundle.yml (hunk 1 above)
Comment (marceltaeumel) [reply .github/workflows/bundle.yml]: Both push and pull_request should have the same filter. Please read my comments above. This is due to being able to try out the builds in my fork. The one who merges this PR still has to do changes to afterwards.

File: .github/workflows/bundle.yml (hunk 2 above)
Comment (marceltaeumel) [reply .github/workflows/bundle.yml]: In the repository settings. Only visible for admins.

File: env_vars (hunk 3 above)
Comment (marceltaeumel) [reply env_vars]: It's just a source file. Same approach as in smalltalkCI. Maybe I should rename it? It's not even "export". Hmm...

File: env_vars (hunk 4 above)
Comment (marceltaeumel) [reply env_vars:4]: There is still support for building Etoys bundles. See http://files.squeak.org/etoys/ . It uses the same base files from files.squeak.org/base. So, you have to replace "Etoys" with "Squeak" when building for "Etoys-trunk" or "Etoys64-trunk".

**Comment by marceltaeumel** [comment]
> Can we also address this issue (duplicate squeak.sh file) or should this go into a separate PR?

Since this is not directly related to supporting GHA, this would be a separate PR.

> As part of this change, could we also consider replacing the recurring "Hi all!" commit by a GitHub action that sends the failure mail to the list (something like this)? I think this would make the commit history look cleaner and it's also more transparent. :-)

Well, GitHub Actions has no support for e-mail notification yet. Using one of the existing mail-send actions, we would need a custom e-mail account for such things. We don't have one at the moment. I would rather not use a personal one for such things.

The recurring "Hi all!" commit is not needed anymore because only Travis had those e-mail notifications, which we used for scheduled runs.

**Comment by LinqLover** [comment]
With a bit of delay, thanks for the explanations! :-)

> > As part of this change, could we also consider replacing the recurring "Hi all!" commit by a GitHub action that sends the failure mail to the list (something like this)? I think this would make the commit history look cleaner and it's also more transparent. :-)
> 
> Well, GitHub Actions has no support for e-mail notification yet. Using one of the existing mail-send actions, we would need a custom e-mail account for such things. We don't have one at the moment. I would rather not use a personal one for such things.

I already have a working dummy account for exactly this purpose (already using this for some other repos). If you are interested, I can DM you the credentials. :-)

> > Can we also address this issue (duplicate squeak.sh file) or should this go into a separate PR?
> 
> Since this is not directly related to supporting GHA, this would be a separate PR.

Just created https://github.com/squeak-smalltalk/squeak-app/issues/23, will do at any time later if no one else anticipates me. :-)

File: .github/workflows/bundle.yml (hunk 1 above)
Comment (LinqLover) [reply .github/workflows/bundle.yml]: I was rather questioning the existence of either `push` or `pull_request` in general. :-) I am usually going with just `push` to any branch. Or wouldn't this work for forks then?

File: env_vars (hunk 3 above)
Comment (LinqLover) [reply env_vars]: I'd vote for renaming it. For instance, I do not get shellcheck or syntax highlighting in VS Code if the file does not have a suffix.

Comment (krono) [reply env_vars]: I think just `env` is a _very_ popular choice.

Comment (LinqLover) [reply env_vars]: Alright, let's measure this! :-)

| glob pattern | hits on public GitHub repositories |
| --- | --: |
| `env` | [2.8k](https://sourcegraph.com/search?q=context:global+file:/env%24+count:100000&patternType=literal) |
| `*.env` | [29.1k](https://sourcegraph.com/search?q=context:global+file:/.*%5C.env%24+count:100000&patternType=literal) |
| `?*env` | [12.5k](https://sourcegraph.com/search?q=context:global+file:/.*%5B%5E.%5Denv%24+count:100000&patternType=literal) |
| `env_*` | [0.635k](https://sourcegraph.com/search?q=context:global+file:/env_%5B%5E./%5D%2B%24+count:100000&patternType=literal) |

So I guess anything but a file name ending in `_env` would be "popular". And VS Code should probably detect just `env` as environment variables, too. :D

Comment (LinqLover) [reply env_vars]: Just because it is hacktober, FYI: https://github.com/mikestead/vscode-dotenv/pull/28 🙃

**Pull request merged**